packages/models-library/src/models_library/user_preferences.py

```diff
@@ -0,0 +1,177 @@
+from collections.abc import Mapping
+from enum import auto
+from typing import Any, ClassVar, TypeAlias
+
+from pydantic import BaseModel, Field
+from pydantic.main import ModelMetaclass
+
+from .services import ServiceKey
+from .utils.enums import StrAutoEnum
+
+IntStr: TypeAlias = int | str
+DictStrAny = dict[str, Any]
+AbstractSetIntStr: TypeAlias = set[IntStr]
+MappingIntStrAny: TypeAlias = Mapping[IntStr, Any]
+
+
+# NOTE: for pyndatic-2 from pydantic._internal.import _model_construction
+# use _model_construction.ModelMetaclass instead!
+
+
+class _AutoRegisterMeta(ModelMetaclass):
+    _registered_user_preference_classes: ClassVar[dict[str, type]] = {}
+
+    def __new__(cls, name, bases, attrs, *args, **kwargs):
+        new_class = super().__new__(cls, name, bases, attrs, *args, **kwargs)
+
+        if name != cls.__name__:
+            if name in cls._registered_user_preference_classes:
+                msg = (
+                    f"Class named '{name}' was already defined at "
+                    f"{cls._registered_user_preference_classes[name]}."
+                    " Please choose a different class name!"
+                )
+                raise TypeError(msg)
+            cls._registered_user_preference_classes[name] = new_class
+
+        return new_class
+
+
+PreferenceName: TypeAlias = str
+PreferenceIdentifier: TypeAlias = str
+
+
+class _ExtendedBaseModel(BaseModel, metaclass=_AutoRegisterMeta):
+    ...
+
+
+class PreferenceType(StrAutoEnum):
+    BACKEND = auto()
+    FRONTEND = auto()
+    USER_SERVICE = auto()
+
+
+class NoPreferenceFoundError(RuntimeError):
+    ...
+
+
+class _BaseUserPreferenceModel(_ExtendedBaseModel):
+    preference_type: PreferenceType = Field(
+        ..., description="distinguish between the types of preferences"
+    )
+
+    value: Any = Field(
+        ...,
+        description="the value of the preference. Stored as is and cannot be queried over",
```

Comment:
I am not sure I understand the second part of the description "cannot be queried over". is that not exactly the point of getting the preferences? are you sure that is correct?